src/components/c-help-redirect.ts

```diff
@@ -0,0 +1,258 @@
+    "<cstdio>",
+];
+
+const maybe_c_keywords = [
+    "printf",
+    "scanf",
+    "<stdio.h>",
+    "<stdlib.h>",
+    "<math.h>",
+    "malloc",
+    "calloc",
+    "realloc",
+    "free(", //too common to include just "free"
+];
+
+/**
+ * Checks for cpp code in #c-help-text and suggests #cpp-help-text instead, and vice versa
+ */
+export default class CHelpRedirect extends BotComponent {
+    constructor(wheatley: Wheatley) {
+        super(wheatley);
+
+        this.add_command(
+            new TextBasedCommandBuilder("not-c")
+                .set_description("Mark C++ code in the C help channel")
+                .add_user_option({
+                    title: "user",
+                    description: "User who posted the code",
+                    required: false,
+                })
+                .set_handler(this.not_c.bind(this)),
+        );
+
+        this.add_command(
+            new TextBasedCommandBuilder("not-cpp")
+                .set_description("Mark C code in the C++ help channel")
+                .add_user_option({
+                    title: "user",
+                    description: "User who posted the code",
+                    required: false,
+                })
+                .set_handler(this.not_cpp.bind(this)),
+        );
+    }
+
+    async is_not_new_member(message: Discord.Message) {
+        let member: Discord.GuildMember;
+        if (message.member == null) {
+            try {
+                member = await message.guild!.members.fetch(message.author.id);
+            } catch (error) {
+                M.warn("Failed to get user", message.author.id);
+                return false;
+            }
+        } else {
+            member = message.member;
+        }
+        assert(member.joinedTimestamp != null);
+        return Date.now() - member.joinedTimestamp >= 7 * DAY;
+    }
+
+    check_message_for_cpp_code(message: Discord.Message): boolean {
+        if (!message.content.includes(code_block_start)) {
+            // to avoid false positives, only check inside code blocks
+            return false;
+        }
+
+        let text = message.content;
+
+        while (text.search(code_block_start) > -1) {
+            const start = text.search(code_block_start);
+            const end = text.substring(start + code_block_start.length).search(code_block_start);
+            const block = text.substring(start + code_block_start.length, start + code_block_start.length + end);
+
+            for (const keyword of cpp_keywords) {
+                if (block.includes(keyword)) {
+                    return true;
+                }
+            }
+
+            text = text.substring(start + code_block_start.length + end + code_block_start.length);
+        }
+        return false;
+    }
+
+    check_message_for_c_code(message: Discord.Message): boolean {
+        if (!message.content.includes(code_block_start)) {
+            // to avoid false positives, only check inside code blocks
```

Comment:
Most of the people asking questions in the wrong channel are beginners who don't know how to type code blocks. It would probably be good to look for things like `cout <<` outside of code blocks. I'd be happy to merge the module with the current approach, though, and the logic can be expanded later.

Reply:
I was aiming to make the automatic check be pretty certain before firing. Nobody wants a bot that spams people with false positives. The gaps can be filled using the manual commands.
I am interested in expanding the checks, when we see how the initial version works in practice. I would also like to see what kind of messages get flagged with the manual commands, and maybe base the future checks on them.

Reply:
Sounds good